### PR TITLE
Added method to create a mock SparkContext, 

### PR DESCRIPTION
**Pull request by smarthi**
added tests for better code coverage of the KMeans Eval code

### CODE REVIEW

File: oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/AbstractKMeansEvalTest.java [hunk 1]
```diff
@@ -20,6 +20,14 @@
 import java.util.List;
 import java.util.Map;
 
+import org.apache.spark.SparkConf;
```
Comment (srowen) [inline oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/AbstractKMeansEvalTest.java]: I'd rather put all this in an integration test, an "IT" class

File: oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/AbstractKMeansEvalTest.java [hunk 2]
```diff
@@ -38,6 +46,18 @@
           {-2.0, 0.0}, {-0.5, -1.0}, {-0.5, 1.0}
       });
   protected static final DistanceFn<double[]> DISTANCE_FN = new SquaredDistanceFn();
+  protected static JavaSparkContext javaSparkContext;
```
Comment (srowen) [inline oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/AbstractKMeansEvalTest.java]: private and non-static?

File: oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/AbstractKMeansEvalTest.java [hunk 3]
```diff
@@ -84,4 +104,23 @@
 
     return new Tuple2<>(minCluster, closestDist);
   }
+
+  private static JavaSparkContext getSparkContext() {
```
Comment (srowen) [inline oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/AbstractKMeansEvalTest.java]: Maybe make this into lazy initialization and make it protected non-static?

File: oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/AbstractKMeansEvalIT.java [hunk 4]
```diff
@@ -38,6 +46,20 @@
           {-2.0, 0.0}, {-0.5, -1.0}, {-0.5, 1.0}
       });
   protected static final DistanceFn<double[]> DISTANCE_FN = new SquaredDistanceFn();
+  private JavaSparkContext javaSparkContext;
+  private static JavaRDD<double[]> pointsRDD;
```
Comment (srowen) [inline oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/AbstractKMeansEvalIT.java]: This is also supposed to be non-static or it's going to clash with other tests running at the same time

File: oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/AbstractKMeansEvalIT.java [hunk 5]
```diff
@@ -27,17 +32,17 @@
 import com.cloudera.oryx.app.kmeans.KMeansPMMLUtils;
 import com.cloudera.oryx.app.kmeans.KMeansPMMLUtilsTest;
 import com.cloudera.oryx.app.kmeans.SquaredDistanceFn;
-import com.cloudera.oryx.common.OryxTest;
+import com.cloudera.oryx.lambda.AbstractSparkIT;
 
-public abstract class AbstractKMeansEvalTest extends OryxTest {
+public abstract class AbstractKMeansEvalIT extends AbstractSparkIT implements Serializable {
```
Comment (srowen) [inline oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/AbstractKMeansEvalIT.java]: The test should not have to implement `Serializable` or really we should fix up the code so the test object is not in a closure.

File: oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/AbstractKMeansEvalIT.java [hunk 6]
```diff
@@ -27,17 +32,17 @@
 import com.cloudera.oryx.app.kmeans.KMeansPMMLUtils;
 import com.cloudera.oryx.app.kmeans.KMeansPMMLUtilsTest;
 import com.cloudera.oryx.app.kmeans.SquaredDistanceFn;
-import com.cloudera.oryx.common.OryxTest;
+import com.cloudera.oryx.lambda.AbstractSparkIT;
 
-public abstract class AbstractKMeansEvalTest extends OryxTest {
```
Comment (srowen) [inline oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/AbstractKMeansEvalIT.java]: Eh, this is still removing the unit tests right? the unit tests seemed fine. I think we're looking at adding an IT for SparkContext-based tests separately

File: oryx-lambda/src/test/java/com/cloudera/oryx/lambda/AbstractSparkIT.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    javaSparkContext = getSparkContext();
+  }
+
+  @After
+  public void tearDown() {
+    javaSparkContext.close();
+    javaSparkContext = null;
+  }
+
+  private JavaSparkContext getSparkContext() {
```
Comment (srowen) [inline oryx-lambda/src/test/java/com/cloudera/oryx/lambda/AbstractSparkIT.java]: Just inline this in `setUp()`; I think it's clearer

File: oryx-lambda/src/test/java/com/cloudera/oryx/lambda/AbstractSparkIT.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+package com.cloudera.oryx.lambda;
+
+import org.apache.spark.SparkConf;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.junit.After;
+import org.junit.Before;
+
+import com.cloudera.oryx.common.OryxTest;
+
+public class AbstractSparkIT extends OryxTest {
```
Comment (srowen) [inline oryx-lambda/src/test/java/com/cloudera/oryx/lambda/AbstractSparkIT.java]: abstract?

File: oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/AbstractKMeansEvalIT.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.List;
+
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.function.Function;
+import org.apache.spark.mllib.linalg.Vector;
+import org.apache.spark.mllib.linalg.Vectors;
+
+import com.cloudera.oryx.lambda.AbstractSparkIT;
+
+public class AbstractKMeansEvalIT extends AbstractSparkIT implements Serializable {
```
Comment (srowen) [inline oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/AbstractKMeansEvalIT.java]: class can be abstract; method can be final.

File: oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/AbstractKMeansEvalIT.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+  protected JavaRDD<Vector> getRddOfVectors(List<double[]> pointsList) {
+    return getJavaSparkContext().parallelize(pointsList).map(
+        new Function<double[], Vector>() {
```
Comment (srowen) [inline oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/AbstractKMeansEvalIT.java]: This is likely the problem with `Serializable`. This anonymous inner class is non-static. Break it out as a private static member

File: oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/AbstractKMeansEvalTest.java [hunk 11]
```diff
@@ -27,17 +27,16 @@
 import com.cloudera.oryx.app.kmeans.KMeansPMMLUtils;
 import com.cloudera.oryx.app.kmeans.KMeansPMMLUtilsTest;
 import com.cloudera.oryx.app.kmeans.SquaredDistanceFn;
-import com.cloudera.oryx.common.OryxTest;
 
-public abstract class AbstractKMeansEvalTest extends OryxTest {
```
Comment (srowen) [inline oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/AbstractKMeansEvalTest.java]: This is still changing the unit test. I really think they can be left alone. We are adding a new test class only.

File: oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/KMeansEvalIT.java [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.mllib.linalg.Vector;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.cloudera.oryx.app.kmeans.ClusterInfo;
+
+public final class KMeansEvalIT extends AbstractKMeansEvalIT {
```
Comment (srowen) [inline oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/KMeansEvalIT.java]: OK, but you are still duplicating the tests. I'm not against it, but it seems redundant. I don't see the need for a second abstract superclass here. There's just one subclass, that I thought might only test things like `testFetchSampleEvalData`. I don't see a test of the sampling?

Comment (smarthi) [inline oryx-app-mllib/src/test/java/com/cloudera/oryx/app/mllib/kmeans/KMeansEvalIT.java]: The first test is a sampling one, will remove the other redundant tests.

**Comment by srowen** [comment]
Hm, now the unit tests are deleted in favor of integration tests only. There are really a lot of ways to do this and all are fine. I had wanted to keep this simple and leave the unit tests as they are since they were fine as unit tests. But I thought we had both wanted to just add a new IT for the RDD-related functionality and leave unit tests exactly as they are. Maybe I can just make this change rather than go around again?

**Comment by smarthi** [comment]
There's better code coverage with having those as IT then if they were unit tests. That was my thinking. I am open to having it either way, I can restore the previous unit test back and just have a sampling test in IT .

**Comment by srowen** [comment]
OK yeah, let's go with this. I like that it's all back in one place
actually. It's good. Ill merge.

**Pull request merged**